cavern/src/main/java/org/opencadc/cavern/files/PutAction.java

```diff
@@ -227,8 +224,27 @@ public void doAction() throws Exception {
                 throw e;
             }
         } finally {
-            log.debug("put: done " + nodeURI.getURI().toASCIIString());
+            if (successful) {
+                log.debug("put: done " + nodeURI.getURI().toASCIIString());
+            } else if (putStarted) {
+                cleanUpOnFailure(target, node, rootPath);
+            }
+        }
+    }
+    
+    private void cleanUpOnFailure(Path target, Node node, Path rootPath) throws IOException {
+        log.debug("clean up on put failure " + target);
+        Files.delete(target);
+        // restore empty DataNode: remove props that are no longer applicable
+        Iterator<NodeProperty> i = node.getProperties().iterator();
```

Comment:
The class Node.java has a method getPropertyValue() you can use so that you don't need to iterate through the properties.

Reply:
Thanks for pointing that out. Will use 
`public NodeProperty findProperty(String propertyURI) `
which seems more appropriate for the occasion.